src/lib/pure_pursuit/PurePursuit.cpp

```diff
@@ -53,43 +54,85 @@ void PurePursuit::updateParams()
 
 }
 
-float PurePursuit::calcDesiredHeading(const Vector2f &curr_wp_ned, const Vector2f &prev_wp_ned,
-				      const Vector2f &curr_pos_ned,
-				      const float vehicle_speed)
+float PurePursuit::updatePurePursuit(const Vector2f &curr_wp_ned, const Vector2f &prev_wp_ned,
+				     const Vector2f &curr_pos_ned, const float vehicle_speed)
+{
+	_target_bearing = calcTargetBearing(curr_wp_ned, prev_wp_ned, curr_pos_ned, vehicle_speed);
+
+	if (PX4_ISFINITE(_target_bearing)) {
+		publishPurePursuit();
+	}
+
+	return _target_bearing;
+}
+
+float PurePursuit::calcTargetBearing(const Vector2f &curr_wp_ned, const Vector2f &prev_wp_ned,
+				     const Vector2f &curr_pos_ned, const float vehicle_speed)
 {
 	// Check input validity
-	if (!curr_wp_ned.isAllFinite() || !curr_pos_ned.isAllFinite() || vehicle_speed < -FLT_EPSILON
-	    || !PX4_ISFINITE(vehicle_speed) || !prev_wp_ned.isAllFinite()) {
+	if (!curr_wp_ned.isAllFinite() || !curr_pos_ned.isAllFinite() || !PX4_ISFINITE(vehicle_speed)
+	    || !prev_wp_ned.isAllFinite()) {
 		return NAN;
 	}
 
-	_lookahead_distance = math::constrain(_params.lookahead_gain * vehicle_speed,
-					      _params.lookahead_min, _params.lookahead_max);
-
 	// Pure pursuit
-	const Vector2f curr_pos_to_curr_wp = curr_wp_ned - curr_pos_ned;
+	_lookahead_distance = math::constrain(_params.lookahead_gain * fabsf(vehicle_speed),
+					      _params.lookahead_min, _params.lookahead_max);
+	_curr_pos_to_curr_wp = curr_wp_ned - curr_pos_ned;
 	const Vector2f prev_wp_to_curr_wp = curr_wp_ned - prev_wp_ned;
+	const Vector2f prev_wp_to_curr_pos = curr_pos_ned - prev_wp_ned;
+	const Vector2f prev_wp_to_curr_wp_u = prev_wp_to_curr_wp.unit_or_zero();
+	_distance_along_path = (prev_wp_to_curr_pos * prev_wp_to_curr_wp_u) *
+			       prev_wp_to_curr_wp_u; // Projection of prev_wp_to_curr_pos onto prev_wp_to_curr_wp
+	const Vector2f curr_pos_to_path = _distance_along_path -
+					  prev_wp_to_curr_pos; // Shortest vector from the current position to the path
+	_crosstrack_error = sign(prev_wp_to_curr_wp(1) * curr_pos_to_path(
```

Comment:
I think this can use just sign(curr_pos_to_path.dot(curr_pos_to_curr_wp))*curr_pos_to_path.norm()
Dot product can be used to measure an angle, between two vectors. But I reckon yours should also work.